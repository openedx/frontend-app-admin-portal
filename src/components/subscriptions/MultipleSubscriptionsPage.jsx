import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import {
  Row,
  Col,
  Card,
  Hyperlink,
} from '@edx/paragon';
import { connect } from 'react-redux';

import LoadingMessage from '../LoadingMessage';
import { SubscriptionContext } from './SubscriptionData';
import SubscriptionExpiration from './expiration/SubscriptionExpiration';
import MultipleSubscriptionsPicker from './MultipleSubscriptionPicker';
import {
  DEFAULT_LEAD_TEXT,
} from './data/constants';
import { sortSubscriptionsByStatus } from './data/utils';
import { ROUTE_NAMES } from '../EnterpriseApp/constants';
import { configuration } from '../../config';

const MultipleSubscriptionsPage = ({
  enterpriseSlug,
  redirectPage,
  leadText,
  createActions,
}) => {
  const { loading, data } = useContext(SubscriptionContext);
  const subscriptions = data.results;

  if (loading) {
    return <LoadingMessage className="subscriptions" />;
  }

  if (subscriptions.length === 0) {
    return (
      <Card>
        <Card.Section className="text-center">
          <Row>
            <Col xs={12} lg={{ span: 8, offset: 2 }}>
              <h3 className="mb-3">No subscription plans for your organization</h3>
              <p>
                We were unable to find any active subscription plans for your organization. Please contact
                Customer Support if you have questions.
              </p>
              <Hyperlink
                className="btn btn-brand"
                target="_blank"
                destination={configuration.ENTERPRISE_SUPPORT_URL}
              >
                Contact support
              </Hyperlink>
            </Col>
          </Row>
        </Card.Section>
      </Card>
    );
  }

  if (subscriptions.length === 1) {
    return (
      <Redirect to={`/${enterpriseSlug}/admin/${redirectPage}/${subscriptions[0].uuid}`} />
    );
  }

  const sortedSubscriptions = sortSubscriptionsByStatus(subscriptions);

  return (
    <>
      <SubscriptionExpiration />
      <MultipleSubscriptionsPicker
        enterpriseSlug={enterpriseSlug}
        leadText={leadText}
        subscriptions={sortedSubscriptions}
        redirectPage={redirectPage}
        createActions={createActions}
      />
    </>
  );
};

MultipleSubscriptionsPage.defaultProps = {
  redirectPage: `${ROUTE_NAMES.subscriptionManagement}`,
  leadText: DEFAULT_LEAD_TEXT,
  createActions: null,
};

MultipleSubscriptionsPage.propTypes = {
  enterpriseSlug: PropTypes.string.isRequired,
  redirectPage: PropTypes.string,
  leadText: PropTypes.string,
  createActions: PropTypes.func,
};

const mapStateToProps = (state) => ({
  enterpriseSlug: state.portalConfiguration.enterpriseSlug,
});

export default connect(mapStateToProps)(MultipleSubscriptionsPage);
