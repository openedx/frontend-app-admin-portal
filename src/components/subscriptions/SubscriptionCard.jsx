import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Link } from 'react-router-dom';
import {
  Card,
  Badge,
  Button,
  Stack,
  Row,
  Col,
} from '@edx/paragon';

import classNames from 'classnames';
import { getSubscriptionStatus } from './data/utils';
import { ACTIVE, SCHEDULED, SUBSCRIPTION_STATUS_BADGE_MAP } from './data/constants';

const SubscriptionCard = ({
  subscription,
  createActions,
}) => {
  const {
    title,
    startDate,
    expirationDate,
    licenses = {},
  } = subscription;

  const formattedStartDate = moment(startDate).format('MMMM D, YYYY');
  const formattedExpirationDate = moment(expirationDate).format('MMMM D, YYYY');
  const subscriptionStatus = getSubscriptionStatus(subscription);

  const renderDaysUntilPlanStartText = (className) => {
    if (!(subscriptionStatus === SCHEDULED)) {
      return null;
    }

    const now = moment();
    const planStart = moment(startDate);
    const daysUntilPlanStart = planStart.diff(now, 'days');
    const hoursUntilPlanStart = planStart.diff(now, 'hours');

    return (
      <span className={classNames('d-block small', className)}>
        Plan begins in {
          daysUntilPlanStart > 0 ? `${daysUntilPlanStart} day${daysUntilPlanStart > 1 ? 's' : ''}`
            : `${hoursUntilPlanStart} hour${hoursUntilPlanStart > 1 ? 's' : ''}`
        }
      </span>
    );
  };

  const renderActions = () => {
    const actions = createActions?.(subscription) || [];

    if (actions.length === 0) {
      return null;
    }

    return actions.map(action => (
      <Button
        key={action.to}
        variant={action.variant}
        as={Link}
        to={action.to}
      >
        {action.buttonText}
      </Button>
    ));
  };

  const renderCardHeader = () => {
    const subtitle = (
      <div className="d-flex flex-wrap align-items-center">
        <Stack direction="horizontal" gap={3}>
          <Badge variant={SUBSCRIPTION_STATUS_BADGE_MAP[subscriptionStatus].variant}>
            {subscriptionStatus}
          </Badge>
          <span>
            {formattedStartDate} - {formattedExpirationDate}
          </span>
        </Stack>
      </div>
    );

    return (
      <Card.Header
        className={classNames({
          'pb-1': subscriptionStatus !== ACTIVE,
        })}
        title={title}
        subtitle={subtitle}
        actions={(
          <div>
            {renderActions() || renderDaysUntilPlanStartText('mt-4')}
          </div>
        )}
      />
    );
  };

  const renderCardSection = () => {
    if (subscriptionStatus !== ACTIVE) {
      return null;
    }

    return (
      <Card.Section
        title="Licenses"
        muted
      >
        <Row className="d-flex flex-row justify-content-between w-md-75">
          {['Assigned', 'Activated', 'Allocated', 'Unassigned'].map(licenseStatus => (
            <Col xs="6" md="auto" className="d-flex flex-column mb-3 mb-md-0" key={licenseStatus}>
              <span className="small">{licenseStatus}</span>
              <span>{licenses[licenseStatus.toLowerCase()]} of {licenses.total}</span>
            </Col>
          ))}
        </Row>
      </Card.Section>
    );
  };

  return (
    <Card
      orientation="horizontal"
      className={classNames('subscription-card', {
        'pb-4': subscriptionStatus !== ACTIVE,
      })}
    >
      <Card.Body>
        <Stack gap={4}>
          {renderCardHeader()}
          {renderCardSection()}
        </Stack>
      </Card.Body>
    </Card>
  );
};

SubscriptionCard.defaultProps = {
  createActions: null,
};

SubscriptionCard.propTypes = {
  subscription: PropTypes.shape({
    startDate: PropTypes.string.isRequired,
    expirationDate: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    licenses: PropTypes.shape({
      assigned: PropTypes.number.isRequired,
      activated: PropTypes.number.isRequired,
      allocated: PropTypes.number.isRequired,
      unassigned: PropTypes.number.isRequired,
      total: PropTypes.number.isRequired,
    }),
  }).isRequired,
  createActions: PropTypes.func,
};

export default SubscriptionCard;
