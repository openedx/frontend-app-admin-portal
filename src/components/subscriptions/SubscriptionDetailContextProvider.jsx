import React, {
  createContext, useContext, useMemo, useState,
} from 'react';
import PropTypes from 'prop-types';
import {
  DEFAULT_PAGE, ACTIVATED, REVOKED, ASSIGNED,
} from './data/constants';
import { useSubscriptionUsersOverview, useSubscriptionUsers } from './data/hooks';
import { SubscriptionContext } from './SubscriptionData';

export const SubscriptionDetailContext = createContext({});
export const defaultStatusFilter = [ASSIGNED, ACTIVATED, REVOKED].join();

const SubscriptionDetailContextProvider = ({
  children, subscription, disableDataFetching,
}) => {
  // Initialize state needed for the subscription detail view and provide in SubscriptionDetailContext
  const { data: subscriptions, errors, setErrors } = useContext(SubscriptionContext);
  const hasMultipleSubscriptions = subscriptions.count > 1;
  const [currentPage, setCurrentPage] = useState(DEFAULT_PAGE);
  const [searchQuery, setSearchQuery] = useState(null);
  const [overview, forceRefreshOverview] = useSubscriptionUsersOverview({
    subscriptionUUID: subscription.uuid,
    search: searchQuery,
    errors,
    setErrors,
    isDisabled: disableDataFetching,
  });
  const [userStatusFilter, setUserStatusFilter] = useState(defaultStatusFilter);

  const [users, forceRefreshUsers, loadingUsers] = useSubscriptionUsers({
    currentPage,
    searchQuery,
    subscriptionUUID: subscription.uuid,
    errors,
    setErrors,
    userStatusFilter,
    isDisabled: disableDataFetching,
  });

  const forceRefreshDetailView = () => {
    forceRefreshUsers();
    forceRefreshOverview();
  };

  const context = useMemo(() => ({
    currentPage,
    hasMultipleSubscriptions,
    forceRefreshOverview,
    overview,
    searchQuery,
    setCurrentPage,
    setSearchQuery,
    subscription,
    users,
    forceRefreshUsers,
    loadingUsers,
    setUserStatusFilter,
    forceRefreshDetailView,
  }), [
    currentPage,
    userStatusFilter,
    searchQuery,
    hasMultipleSubscriptions,
    overview,
    subscription,
    users,
    loadingUsers,
  ]);
  return (
    <SubscriptionDetailContext.Provider value={context}>
      {children}
    </SubscriptionDetailContext.Provider>
  );
};

SubscriptionDetailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
  subscription: PropTypes.shape({
    uuid: PropTypes.string.isRequired,
  }).isRequired,
  disableDataFetching: PropTypes.bool,
};

SubscriptionDetailContextProvider.defaultProps = {
  disableDataFetching: false,
};

export default SubscriptionDetailContextProvider;
