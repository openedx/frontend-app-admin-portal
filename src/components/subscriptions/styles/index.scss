.manage-subscription {
    display: flex;
    flex-direction: column;
    flex: 1;

    .add-users-dropdown {
        .dropdown-menu {
            z-index: $zindex-fixed;
        }
    }

    .subscription-card {
        margin-bottom: map_get($spacers, 4);
    }

    @include media-breakpoint-up(lg) {
        .add-users-dropdown {
            float: right;

            .dropdown-menu {
                left: inherit;
                right: 0;
            }
        }
    }
}

.pgn__modal-title .enroll-header {
    display: inline-block;
    background-color: transparent;
    color: $red;
}
