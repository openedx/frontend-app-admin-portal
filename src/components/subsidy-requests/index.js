import SubsidyRequestsContextProvider, { SubsidyRequestsContext } from './SubsidyRequestsContextProvider';

export default SubsidyRequestsContextProvider;
export { SubsidyRequestsContext };
