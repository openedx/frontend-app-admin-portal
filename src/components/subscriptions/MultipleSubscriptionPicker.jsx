import React from 'react';
import PropTypes from 'prop-types';
import {
  Row,
  Col,
} from '@edx/paragon';

import SubscriptionCard from './SubscriptionCard';
import { DEFAULT_LEAD_TEXT } from './data/constants';

const MultipleSubscriptionsPicker = ({
  leadText, subscriptions, createActions,
}) => (
  <>
    <Row>
      <Col xs="12">
        <h2>Plans</h2>
        <p>{leadText}</p>
      </Col>
      <Col lg="10">
        {subscriptions.map(subscription => (
          <SubscriptionCard
            key={subscription.uuid}
            subscription={subscription}
            createActions={createActions}
          />
        ))}
      </Col>
    </Row>
  </>
);

MultipleSubscriptionsPicker.defaultProps = {
  leadText: DEFAULT_LEAD_TEXT,
  createActions: null,
};

MultipleSubscriptionsPicker.propTypes = {
  leadText: PropTypes.string,
  subscriptions: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  createActions: PropTypes.func,
};

export default MultipleSubscriptionsPicker;
