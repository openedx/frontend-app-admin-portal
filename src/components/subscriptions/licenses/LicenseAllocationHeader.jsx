import React, { useContext } from 'react';
import { SubscriptionDetailContext } from '../SubscriptionDetailContextProvider';
import { SubsidyRequestsContext } from '../../subsidy-requests';
import NewFeatureAlertBrowseAndRequest from '../../NewFeatureAlertBrowseAndRequest';
import { SUPPORTED_SUBSIDY_TYPES } from '../../../data/constants/subsidyRequests';

const LicenseAllocationHeader = () => {
  const {
    subscription,
  } = useContext(SubscriptionDetailContext);
  const { subsidyRequestConfiguration } = useContext(SubsidyRequestsContext);

  // don't show alert if the enterprise already has subsidy requests enabled
  const isBrowseAndRequestFeatureAlertShown = subsidyRequestConfiguration?.subsidyType
    === SUPPORTED_SUBSIDY_TYPES.license && !subsidyRequestConfiguration?.subsidyRequestsEnabled;

  return (
    <>
      {isBrowseAndRequestFeatureAlertShown && <NewFeatureAlertBrowseAndRequest />}
      <h2 className="mb-2">License Allocation</h2>
      <p className="lead">
        {subscription.licenses?.allocated}
        {' of '}
        {subscription.licenses?.total} licenses allocated
      </p>
    </>
  );
};

export default LicenseAllocationHeader;
