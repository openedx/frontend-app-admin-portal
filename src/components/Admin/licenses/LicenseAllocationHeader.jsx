import React, { useContext } from 'react';
import { Badge } from '@edx/paragon';
import { SubscriptionDetailContext } from '../../subscriptions/SubscriptionDetailContextProvider';
import { SubsidyRequestsContext } from '../../subsidy-requests';
import NewFeatureAlertBrowseAndRequest from '../../NewFeatureAlertBrowseAndRequest';
import { SUPPORTED_SUBSIDY_TYPES } from '../../../data/constants/subsidyRequests';

const LicenseAllocationHeader = () => {
  const {
    subscription,
  } = useContext(SubscriptionDetailContext);
  const { subsidyRequestConfiguration } = useContext(SubsidyRequestsContext);

  // don't show alert if the enterprise already has subsidy requests enabled
  const isBrowseAndRequestFeatureAlertShown = subsidyRequestConfiguration?.subsidyType
    === SUPPORTED_SUBSIDY_TYPES.license && !subsidyRequestConfiguration?.subsidyRequestsEnabled;
  return (
    <>
      {isBrowseAndRequestFeatureAlertShown && <NewFeatureAlertBrowseAndRequest />}
      <div className="ml-4">
        <h4 style={{ display: 'inline-block' }}>Licenses:</h4>
        <Badge className="mr-2 ml-2" variant="light">Unassigned: {subscription.licenses?.unassigned}
          {' of '}
          {subscription.licenses?.total} total
        </Badge>
        <Badge variant="light">Activated: {subscription.licenses?.activated}
          {' of '}
          {subscription.licenses?.assigned} assigned
        </Badge>
      </div>
    </>
  );
};

export default LicenseAllocationHeader;
