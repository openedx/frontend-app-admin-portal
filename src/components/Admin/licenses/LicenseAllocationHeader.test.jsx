import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import LicenseAllocationHeader from './LicenseAllocationHeader';
import { SubscriptionDetailContext } from '../../subscriptions/SubscriptionDetailContextProvider';
import { SubsidyRequestsContext } from '../../subsidy-requests';

// eslint-disable-next-line react/jsx-no-constructed-context-values
describe('LicenseAllocationHeader', () => {
  const mockStore = configureMockStore();

  // eslint-disable-next-line react/prop-types
  const SubscriptionDetailContextWrapper = ({ children }) => (
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    <SubscriptionDetailContext.Provider value={{
      subscription: {
        licenses: {
          unassigned: 1,
          total: 2,
          activated: 1,
          assigned: 1,
        },
      },
    }}
    >
      {children}
    </SubscriptionDetailContext.Provider>
  );

  // eslint-disable-next-line react/prop-types
  const SubsidyRequestsContextWrapper = ({ children }) => (
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    <SubsidyRequestsContext.Provider value={{
      subsidyRequests: {
        count: 1,
        results: [
          {
            uuid: '123',
            status: 'approved',
            enterprise_customer_name: 'Test Enterprise',
            enterprise_customer_uuid: '456',
            num_licenses: 1,
            start_date: '2020-01-01',
            expiration_date: '2020-01-01',
            subscription_plan_title: 'Test Plan',
            subscription_plan_uuid: '789',
            user_email: 'edx@example.com',
            user_account_uuid: '101112',
          },
        ],
      },
      subsidyRequestConfiguration: {
        subsidyType: 'license',
        subsidyRequestsEnabled: true,
      },
    }}
    >
      {children}
    </SubsidyRequestsContext.Provider>
  );

  const LicenseAllocationHeaderWrapper = () => (
    <Provider store={mockStore({})}>
      <MemoryRouter>
        <SubscriptionDetailContextWrapper>
          <SubsidyRequestsContextWrapper>
            <LicenseAllocationHeader />
          </SubsidyRequestsContextWrapper>
        </SubscriptionDetailContextWrapper>
      </MemoryRouter>
    </Provider>
  );

  it('renders without crashing', () => {
    const tree = renderer.create(<LicenseAllocationHeaderWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
