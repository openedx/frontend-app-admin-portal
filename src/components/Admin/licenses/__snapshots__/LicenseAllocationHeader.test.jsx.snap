// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseAllocationHeader renders without crashing 1`] = `
<div
  className="ml-4"
>
  <h4
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    Licenses:
  </h4>
  <span
    className="mr-2 ml-2 badge badge-light"
  >
    Unassigned: 
    1
     of 
    2
     total
  </span>
  <span
    className="badge badge-light"
  >
    Activated: 
    1
     of 
    2
     assigned
  </span>
</div>
`;
